scipy/sparse/sputils.py

```diff
@@ -189,12 +190,20 @@ def isintlike(x):
     """Is x appropriate as an index into a sparse matrix? Returns True
     if it can be cast safely to a machine int.
     """
-    if not isscalarlike(x):
+    if np.ndim(x) != 0:
         return False
     try:
-        return bool(int(x) == x)
+        operator.index(x)
     except (TypeError, ValueError):
-        return False
+        try:
+            loose_int = bool(int(x) == x)
+        except (TypeError, ValueError):
+            return False
+        if loose_int:
+            warnings.warn("Inexact indices into sparse matrices are deprecated",
+                          DeprecationWarning)
```

Comment:
Shouldn't the DeprecationWarning not be raised unconditionally if operator.index fails?

Reply:
No, the warning should only appear for places where `isintlike(x)` has the "wrong" value. If the old conversion gives the same result as `operator.index`, there's nothing to warn about - the `False` already gets turned into an error down the line,